src/Runner.Worker/StepsRunner.cs

```diff
@@ -76,10 +76,33 @@ public async Task RunAsync(IExecutionContext jobContext)
                 // Start
                 step.ExecutionContext.Start();
 
-                // Set GITHUB_ACTION
+                // Populate env context for each step
+                Trace.Info("Initialize Env context for step");
+#if OS_WINDOWS
+                var envContext = new DictionaryContextData();
+#else
+                var envContext = new CaseSensitiveDictionaryContextData();
+#endif
+                step.ExecutionContext.ExpressionValues["env"] = envContext;
```

Comment:
every step will has a new `env` context copy that include's `env` from job and `::set-env::`